### PR TITLE
Deploy Research Instance (Properly)

### PR DESCRIPTION
**Pull request by chadlagore**
:hourglass: **Status**: _Ready for review_
:tickets: **Ticket(s)**: #18, #24, #23 

## :construction_worker: Changes

+ Wrote a prod shim prod image to layer research environment on top of.
+ Makefile.
+ Jupyter settings.
+ Development image.
+ Travis and CD.

## :flashlight: Testing Instructions

```bash
make       # Builds two images
make dev   # Opens dev up on localhost:8080 (pw in slack)
make run   # Eventually runs the server, for now drops into container.
```

## 🌮 Travis

+ Travis builds and pushes both development and production images when the master branch is updated.
+ We push to dockerhub with a deploy bot (https://hub.docker.com/u/ubclaunchpad/).
+ On digital ocean, we can restart the research server by killing the running container and running `make run` from the `/research` folder.

#### TODO

Add a hook on DO to replace the existing container when new code hits the master branch here.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+services:
+  - docker
+
+script:
+  - make
+
+after_success:
+  - if [ "$TRAVIS_BRANCH" == "master" ]; then
+    docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
```
Comment (iKevinY) [inline .travis.yml]: Can we move the Docker commands into a Makefile target and call that here from within this if-block? It's easier to maintain the Makefile than making changes to `.travis.yml`.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
+FROM ubuntu:16.10
+MAINTAINER Chad Lagore <chad.d.lagore@gmail.com>
+
+# Basic python 3.6 steup.
+RUN apt-get update -qq
+RUN apt-get install -y python3.6 python3-pip python3-pil
+RUN apt-get install -y libpq-dev libjpeg-dev
+RUN pip3 install --upgrade pip
+
+# Add core dependnecies deps.
```
Comment (iKevinY) [inline Dockerfile]: Minor typo: `dependnecies deps` → `dependencies`

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Image names
+PROD_IMAGE_NAME=minutes-prod
+DEV_IMAGE_NAME=minutes-dev
+
+# Ports
+PROD_PORT=80
+DEV_PORT=8080
+
+.PHONY: dev
```
Comment (iKevinY) [inline Makefile]: The other targets should be added to `.PHONY` as well.

File: dev/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Start from base image and add development tooling on top.
+FROM minutes-prod:latest
+
+# Add additional developments.
```
Comment (iKevinY) [inline dev/Dockerfile]: Maybe "add additional development requirements"?

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+nb/*
+
```
Comment (iKevinY) [inline .gitignore]: Extra newline here (you added two in the previous change 😛).

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    docker tag $DEV_IMAGE:$TAG $ORGANIZATION/$DEV_IMAGE:$TAG;
+    docker push $ORGANIZATION/$DEV_IMAGE:$TAG;
+    docker tag $PROD_IMAGE:$TAG $ORGANIZATION/$PROD_IMAGE:$TAG;
+    docker push $ORGANIZATION/$PROD_IMAGE:$TAG;
+    fi;
```
Comment (iKevinY) [inline .travis.yml]: The semicolon after `fi` is unnecessary.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,24 @@
+FROM ubuntu:16.10
+MAINTAINER Chad Lagore <chad.d.lagore@gmail.com>
+
+# Basic python 3.6 steup.
```
Comment (iKevinY) [inline Dockerfile]: Minor typo: `steup` → `setup` (also, "Python" should be capitalized 😁)

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+all: build-prod build-dev
+
+run:
+	docker run -it \
+			-p $(PROD_PORT):$(PROD_PORT) \
+			$(PROD_IMAGE_NAME)
+
+dev:
+	echo "Opening dev environment on localhost port $(DEV_PORT)."
```
Comment (iKevinY) [inline Makefile]: `echo` should probably be `@echo` in order to suppress the echo command itself from being printed.

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+PROD_IMAGE_NAME=minutes-prod
+DEV_IMAGE_NAME=minutes-dev
+ORGANIZATION=ubclaunchpad
+TAG=latest
+
+# Ports
+PROD_PORT=80
+DEV_PORT=8080
+
+.PHONY: dev run build-prod build-dev push-dev push-prod
```
Comment (iKevinY) [inline Makefile]: The `all` target should also be added to `.PHONY`.

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+build-prod:
+	docker build --rm -t $(PROD_IMAGE_NAME) .
+
+build-dev:
+	docker build --rm \
+			-f ./dev/Dockerfile \
+			-t $(DEV_IMAGE_NAME) .
+
+push-dev:
+		docker login -u=$(DOCKER_USERNAME) -p=$(DOCKER_PASSWORD) && \
```
Comment (iKevinY) [inline Makefile]: There's an extra tab here (and also in the first command of the `push-prod` target).

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    	docker tag $(DEV_IMAGE_NAME):$(TAG) $(ORGANIZATION)/$(DEV_IMAGE_NAME):$(TAG) && \
+    	docker push $(ORGANIZATION)/$(DEV_IMAGE_NAME):$(TAG)
+
+push-prod:
+		docker login -u=$(DOCKER_USERNAME) -p=$(DOCKER_PASSWORD) && \
+    	docker tag $(DEV_IMAGE_NAME):$(TAG) $(ORGANIZATION)/$(DEV_IMAGE_NAME):$(TAG) && \
+    	docker push $(ORGANIZATION)/$(DEV_IMAGE_NAME):$(TAG)
```
Comment (iKevinY) [inline Makefile]: Missing a newline at the end of `Makefile`.

**Comment by chadlagore** [comment]
Good call, cleaned up the Makefile to accommodate.

**Comment by chadlagore** [comment]
Still has a bug when the container never existed in the first place hang on 😅

**Comment by chadlagore** [comment]
There, hope thats not too awkward.

```
docker rm -f `docker ps -aq --filter name=$(PROD_IMAGE_NAME)`
```

☝️ That barks when the argument is empty (cant find the container). In all other cases it erases all remnants of the container (as desired).

**Comment by iKevinY** [comment]
Looks good! Now's a good time to standardize how we merge larger PRs into `master`; I typically like to manually squash intermediate commits so that each commit is an atomic change (and there aren't any loose "fixup" commits), but that can admittedly be tedious at times.

**Comment by chadlagore** [comment]
I'm ok with squashing, it can make histories a bit more ambiguous, but I've never had a direct problem with that. Lets squash :+1:.

**Comment by chadlagore** [comment]
Ah, nm I see what you mean. So a PR like this would get reduced to 3 or 4 atomic commits. That is probably good practice.

**Comment by chadlagore** [comment]
I will use that approach on the next PR. It easy to squash say, your last 4 commits into something atomic, but a bit more annoying to go back and squash the pieces of an existing PR into fewer pieces. That said, this entire PR strikes me as an atomic unit and I will squash it.

**Pull request merged**